### PR TITLE
Maintenance Allowed IP list

### PR DESCRIPTION
**Pull request by jaxwilko**
This PR adds support for whitelisting IP addresses to be allowed to visit the frontend while in maintenance mode. This is really useful when you have a site in maintenance mode to avoid being indexed by crawlers or for staging sites where automated tools need to be ran.

Backend Form

Uses `Request::ip()` under the hood for IP detection.

<a href="https://gitpod.io/#https://github.com/wintercms/winter/pull/147"></a>

### CODE REVIEW

File: modules/cms/models/maintenancesetting/fields.yaml [hunk 1]
```diff
@@ -15,4 +15,21 @@ fields:
...
+        form:
+            fields:
+
+                ip:
+                    label: cms::lang.maintenance.ip_whitelist.ip
+                    type: text
+                    span: left
+
+                label:
+                    label: cms::lang.maintenance.ip_whitelist.ip
```
Comment (matteotrubini) [inline modules/cms/models/maintenancesetting/fields.yaml]: `cms::lang.maintenance.ip_whitelist.label` ?

Comment (jaxwilko) [reply modules/cms/models/maintenancesetting/fields.yaml]: Nice spot, totally missed that :+1:

File: modules/cms/models/MaintenanceSetting.php [hunk 2]
```diff
@@ -89,4 +89,15 @@ public function afterFetch()
             $this->is_enabled = false;
         }
     }
+
+    /**
+     * Returns bool on if IP is in the whitelist
+     * @return bool
+     */
```
Comment (LukeTowers) [inline modules/cms/models/MaintenanceSetting.php]:
```suggestion
    /**
     * Check if the provided IP is in the whitelist
     *
     * @param string $ip
     * @return bool
     */
```

File: modules/cms/models/MaintenanceSetting.php [hunk 3]
```diff
@@ -89,4 +89,17 @@ public function afterFetch()
...
+     * Check if the provided IP is in the whitelist
+     *
+     * @param string $ip
+     * @return bool
+     */
+    public static function inWhitelist(string $ip): bool
+    {
+        return in_array($ip, array_map(function ($arr) {
+            return $arr['ip'];
+        }, static::get('ip_whitelist', [])));
```
Comment (LukeTowers) [inline modules/cms/models/MaintenanceSetting.php]:
```suggestion
        return \Symfony\Component\HttpFoundation\IpUtils::checkIp($ip, static::get('ip_whitelist', []))
```

File: modules/cms/models/MaintenanceSetting.php [hunk 4]
```diff
@@ -3,6 +3,8 @@
 use Model;
 use Cms\Classes\Page;
 use Cms\Classes\Theme;
+use Illuminate\Support\Arr;
```
Comment (LukeTowers) [inline modules/cms/models/MaintenanceSetting.php]: Doesn't October have it's own version?

Comment (mjauvin) [reply modules/cms/models/MaintenanceSetting.php]: Yep:
```
<?php namespace Winter\Storm\Support;

use Illuminate\Support\Arr as ArrHelper;

/**
 * Array helper
 *
 * @author Alexey Bobkov, Samuel Georges
 */
class Arr extends ArrHelper
```

Comment (jaxwilko) [reply modules/cms/models/MaintenanceSetting.php]: Yeah my bad, it got auto completed and I didn't check the import. Updated to Winter's version :)

**Comment by damsfx** [comment]
Nice!
I Always use this option with a little bit of code in themes to show "pre openning" or waiting pages to visitors, and at the same time let backend users to use the website in real conditions.

File: modules/cms/lang/en/lang.php [hunk 5]
```diff
@@ -100,6 +100,11 @@
         'is_enabled' => 'Enable maintenance mode',
         'is_enabled_comment' => 'Select the page to show when maintenance mode is activated.',
         'hint' => 'Maintenance mode will display the maintenance page to visitors who are not signed in to the back-end area.',
+        'ip_whitelist' => [
+            'name' => 'IP Whitelist',
+            'ip' => 'IP Address',
+            'label' => 'Rule Label'
```
Comment (LukeTowers) [inline modules/cms/lang/en/lang.php]:
```suggestion
            'label' => 'Description'
```

**Comment by LukeTowers** [comment]
Minor point, but as developers in general we should probably move away from white/black list terminology towards include/exclude instead as it's more clear to newcomers and less controversial. I would name the field "Allowed IP addresses"  with the visual comment "IP addresses that are permitted to view the site while maintenance mode is active".

**Comment by WebVPF** [comment]
How about adding the line "Your ip ..." before the list of allowed ip addresses, where to display the user's current ip address and the button "Add ip to the allowed list" if the ip is not in the allowed list?

And is it possible to do this functionality the same way to access the backend? Store only an array with allowed IP addresses in the `config/cms.php` file.

**Comment by LukeTowers** [comment]
> And is it possible to do this functionality the same way to access the backend? Store only an array with allowed IP addresses in the config/cms.php file.

@WebVPF that could easily be achieved via a plugin

File: modules/cms/models/MaintenanceSetting.php [hunk 6]
```diff
@@ -89,4 +91,15 @@ public function afterFetch()
...
         }
     }
+
+    /**
+     * Check if the provided IP is in the whitelist
+     *
+     * @param string $ip
+     * @return bool
+     */
+    public static function inAllowedIps(string $ip): bool
```
Comment (LukeTowers) [inline modules/cms/models/MaintenanceSetting.php]:
```suggestion
    public static function isAllowedIp(string $ip): bool
```

File: modules/cms/classes/Controller.php [hunk 7]
```diff
@@ -163,6 +163,7 @@ public function run($url = '/')
         if (
             MaintenanceSetting::isConfigured() &&
             MaintenanceSetting::get('is_enabled', false) &&
+            !MaintenanceSetting::inWhitelist(Request::ip()) &&
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]:
```suggestion
            !MaintenanceSetting::isAllowedIp(Request::ip()) &&
```

**Review by LukeTowers — CHANGES_REQUESTED** [review]

File: modules/cms/models/MaintenanceSetting.php [hunk 8]
```diff
@@ -89,4 +91,15 @@ public function afterFetch()
             $this->is_enabled = false;
         }
     }
+
+    /**
+     * Check if the provided IP is in the whitelist
```
Comment (bennothommo) [inline modules/cms/models/MaintenanceSetting.php]:
```suggestion
     * Check if the provided IP is in the allowed IP list.
```

File: modules/cms/models/maintenancesetting/fields.yaml [hunk 9]
```diff
@@ -15,4 +15,22 @@ fields:
...
+        cssClass: checkbox-align
+
+    allowed_ips:
+        label: cms::lang.maintenance.allowed_ips.name
+        type: repeater
+        commentAbove: cms::lang.maintenance.allowed_ips.description
+        cssClass: checkbox-align
+        form:
+            fields:
+
```
Comment (bennothommo) [inline modules/cms/models/maintenancesetting/fields.yaml]:
```suggestion
```

File: modules/cms/models/maintenancesetting/fields.yaml [hunk 10]
```diff
@@ -15,4 +15,22 @@ fields:
...
+                ip:
+                    label: cms::lang.maintenance.allowed_ips.ip
+                    type: text
+                    span: left
+
```
Comment (bennothommo) [inline modules/cms/models/maintenancesetting/fields.yaml]:
```suggestion
```

**Comment by bennothommo** [comment]
Just to note, as I'm having a look into it now, but this functionality does not support trusted proxies currently. It's using the proxy's IP in the lookup for allowed IPs.

**Review by LukeTowers — APPROVED** [review]

**Comment by bennothommo** [comment]
My previous comment can be ignored. It seems to be something intrinsic to Gitpod in not forwarding on my actual IP - it doesn't happen on other proxies.

**Pull request merged**